controllers/componentversion_controller.go

```diff
@@ -84,51 +83,30 @@ func (r *ComponentVersionReconciler) Reconcile(ctx context.Context, req ctrl.Req
 		return
 	}
 
-	patchHelper, err := patch.NewHelper(obj, r.Client)
-	if err != nil {
-		return ctrl.Result{}, fmt.Errorf("failed to create patchhelper: %w", err)
-	}
+	patchHelper := patch.NewSerialPatcher(obj, r.Client)
 
 	// Always attempt to patch the object and status after each reconciliation.
 	defer func() {
-		// Patching has not been set up, or the controller errored earlier.
-		if patchHelper == nil {
-			return
-		}
-
-		// If still reconciling then reconciliation did not succeed, set to ProgressingWithRetry to
-		// indicate that reconciliation will be retried.
-		if conditions.IsReconciling(obj) {
-			reconciling := conditions.Get(obj, meta.ReconcilingCondition)
-			reconciling.Reason = meta.ProgressingWithRetryReason
-			conditions.Set(obj, reconciling)
-			msg := fmt.Sprintf("Reconciliation did not succeed, retrying in %s", obj.GetRequeueAfter())
-			event.New(r.EventRecorder, obj, eventv1.EventSeverityError, msg, nil)
-		}
-
-		// Set status observed generation option if the component is ready.
-		if conditions.IsReady(obj) {
-			obj.Status.ObservedGeneration = obj.Generation
-			msg := fmt.Sprintf("Reconciliation finished, next run in %s", obj.GetRequeueAfter())
-			vid := fmt.Sprintf("%s:%s", obj.Status.ComponentDescriptor.Name, obj.Status.ReconciledVersion)
-			metadata := make(map[string]string)
-			metadata[v1alpha1.GroupVersion.Group+"/component_version"] = vid
-			event.New(r.EventRecorder, obj, eventv1.EventSeverityInfo, msg, metadata)
-		}
-
-		// Update the object.
-		if err := patchHelper.Patch(ctx, obj); err != nil {
-			retErr = errors.Join(retErr, err)
+		if derr := DeferredStatusUpdate(ctx, patchHelper, obj, r.EventRecorder, obj.GetRequeueAfter()); derr != nil {
+			retErr = errors.Join(retErr, derr)
 		}
 	}()
 
-	rreconcile.ProgressiveStatus(false, obj, meta.ProgressingReason, "reconcilation in progress for component: %s", obj.Spec.Component)
+	// Starts the progression by setting ReconcilingCondition.
+	// This will be checked in defer.
+	// Should only be deleted on a success.
+	rreconcile.ProgressiveStatus(false, obj, meta.ProgressingReason, "reconciliation in progress for component: %s", obj.Spec.Component)
 
 	octx, err := r.OCMClient.CreateAuthenticatedOCMContext(ctx, obj)
 	if err != nil {
-		msg := fmt.Sprintf("authentication failed for repository: %s with error: %s", obj.Spec.Repository.URL, err)
-		conditions.MarkFalse(obj, meta.ReadyCondition, v1alpha1.AuthenticatedContextCreationFailedReason, msg)
-		event.New(r.EventRecorder, obj, eventv1.EventSeverityError, msg, nil)
+		// we don't fail here, because all manifests might have been applied at once or the secret
```

Comment:
Also shouldn't this result in a `stalled` state? If we requeue, can it advance without human intervention?

Reply:
Nope. 

```go
		return ctrl.Result{
			RequeueAfter: obj.GetRequeueAfter(),
		}, nil
```

We specifically don't return an error here because the secret and authentication can be applied a bit later and we don't want to fall back into oblivion and having to retrigger or potentially re-create the object because the secret came in a couple second or a minute too late.

Reply:
Hmm, it should probably be stalled until the auth succeeds though.

We should add a predicate that will trigger reconcile when secret gets added if it matches the named secret for auth. 

This would also constantly requeue if the auth fails and we don't want that.

Reply:
Yeah we can do that. And then, this thing here would return `ctrl.Result{}, nil` so it doesn't constantly reconcile. I guess, that would work out nicely.

Reply:
@souleb @phoban01 This should do the trick. :)